### ISSUE TITLE
Storage API version enum

### ISSUE THREAD

**Issue by Jinming-Hu** [Azure/azure-sdk-for-cpp#1056]
Originally brought up by @ahsonkhan in guidelines meeting.

Plans to expose an API version enum for storage services which change the API version sent over the wire (as a “best-effort”) to enable talking to different clouds and private azure stacks, similar to: https://github.com/Azure/azure-sdk-for-net/blob/1637f4b9ed998c40d9ae3fe054de3382b4769150/sdk/storage/Azure.Storage.Blobs/src/BlobClientOptions.cs#L22-L55

**Comment by Jinming-Hu** [comment]
Hi @ahsonkhan , this was originally brought up by you. Storage team wants to know if this is a feature requested by a customer, so that we can properly prioritize it.

Storage team doesn't think of it as a blocker of GA. We can do it after GA and it's not a breaking change.

**Comment by RickWinter** [comment]
@JeffreyRichter @tg-msft can you comment on if the archboard requires this language consistency feature for GA.

**Comment by JeffreyRichter** [comment]
I would say NO as it is non-breaking to add it later and even if we had it today, the enum would have just 1 symbol associated with it so the customer has no real choice anyway.

**Comment by tg-msft** [comment]
Having a good multi service-version story is important across languages.   Even though you're relying on users compiling in the source, I still think you'll hit diamond dependency problems and need to support multiple service-versions from the latest client-version of your library.

You only need to support the latest service-version when you GA.  Storage tends to release pretty frequently though so I would expect you'd need support for another service-version within a few months of GA.  I would recommend you still ship a `ServiceVersion` (or equivalent) enum at GA with a single value that gets defaulted for customers to the latest.  This will allow customers who care about running against a specific service-version to pin if required so they won't be moved forward automatically by a future client-version.

**Comment by JeffreyRichter** [comment]
Ted makes a good point about the pinning - I forgot about that. So, I agree that it would best to have this before GA but I personally would be OK if it was not there by GA. I know there is lots of work to do and time is runnign out - this is lower priority than some other things.

**Comment by tg-msft** [comment]
I agree this isn't a ship blocker and could be added with only a remote possibility of minor semantic breaks.  It would also be a no-op as implemented today (i.e., just add the enum, add the param, and then you're done) so it's more about deciding on the pattern than anything.  You can wait to plumb this value through to every method when you add a second version.

**Comment by Jinming-Hu** [comment]
blocked by https://github.com/Azure/azure-sdk-for-cpp/issues/1393
